docs/NNNN-schema-registry.md

```diff
@@ -0,0 +1,152 @@
+Configuration is done by setting appropriate values in config.json at the top level of the extension's directory. The keys expected are:
+
+extensionName
+: String identifying the extension. Must be "NNNN-schema-registry".
+
+schemaDigestAlgorithm
+: Algorithm used for calculating safe filenames from schema identifiers. "md5" is the default.
+
+Additionally a schema_registry.json file should be present. This contains the keys -
+
+digestAlgorithm
+: Algorithm used for calculating fixity of the manifest and stored schema files.
+
+manifest
+: An OCFL-style manifest of the stored schema files.
+
+schemaMap
+: Provides a convenient mapping of the original schema identifiers to the digest used to identify the schema.
+
+Example files are shown below.
+
+
+### Implementation
+
+In an OCFL root where this extension is configured, tools reading the OCFL objects are not required to be aware of it. However, accessing local copies of schemata is also likely to decrease latency and may provide a small performance benefit. Thus OCFL reading applications relying on schemata access (eg for validation) may wish to consider reading these from the schema registry.
+
+It is important that a system implementing this extension ensures all OCFL objects written are checked for schema references. Where not already registered, the schema is retrieved and registered as described below.
+
+Two possible implementation cases are considered:
+
+* Where an OCFL is being generated as a snapshot for export or backup purposes, it is reasonable to build or check the completeness of the schema registry as each OCFL object is created. Thus the exported root will contain all necessary schemata.
+
+* Where an OCFL provides the persistence layer for a live repository, it is recommended that a background process discovers (or is informed) of new OCFL object/versions. This process can asynchronously process the object contents, retrieve the required schemata and perform necessary registrations with the schema registry.
+
+Since schemata are typically identified by the URL of the specific version referenced, a hashed representation of the identifier is used to create a safe filename for storage. MD5 is likely to be sufficient although other algorithms may be configured. The extension's inentory stores a map of the hashed identifiers, thus allowing easy access to the appropriate schema from unmodified OCFL objects.
+
+### Registering a schema with the extension
+
+* New OCFL objects/versions are inspected for JSON, XML files. 
+* Within these files, references to external schema ($schema / XML-DTD etc) are extracted
+* The local filename is derived using the specified schemaDigestAlgorithm.
+* schemata_inventory.json is checked for this key. If not already present in the schema repository:
+  * A copy of the remote schema is retrieved, and stored in the schemata/ directory with this name.
+  * The schemata_inventory.json is updated:
+    * a ‘manifest’ object entry is created for fixity.
```

Comment:
How is this entry created?

Reply:
This would be the responsibility of any process that writes to the schema registry.

Reply:
What I meant is that nowhere is it described how the `manifest` is structured. There seems to be an implicit assumption that it is similar to the `manifest` block within an OCFL inventory. This structure doesn't bother me as much as the `schemaMap`, but I do wonder if it would not be more useful if it mapped paths to digests instead. For example:

```json
  "manifest": {
    "schemata/40cdd53d9a263e5466b8954d82d23daa": "91da2b...9c8",
    "schemata/95d751340dcdc784fd759dbc7ddb9633": "31f53b...7ff"
  }
```

That is certainly the structure I'd want if I were validating the files. The inventory manifest maps digests to content paths because of its relationship with version state. That is, logical path -> digest; digest -> content path. It's not clear to me what advantage there is to using the same structure here.